### ISSUE TITLE
CI is broken (TestAccDockerService_fullSpec is failed on master branch)

### ISSUE THREAD

**Issue by suzuki-shunsuke** [terraform-providers/terraform-provider-docker#237]
### Terraform Version

CI by Travis CI. `make testacc`

### Affected Resource(s)
Please list the resources as a list, for example:
- docker_service

### Terraform Configuration Files

https://github.com/terraform-providers/terraform-provider-docker/blob/a7f6cc93009c4052f836c279e68ecd4a96ab238d/docker/resource_docker_service_test.go#L224-L401

### Debug Output
Please provider a link to a GitHub Gist containing the complete debug output: https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

https://travis-ci.org/terraform-providers/terraform-provider-docker/jobs/635539888?utm_medium=notification&utm_source=github_status

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
What should have happened?

CI should be passed.

### Actual Behavior
What actually happened?

CI fails because `TestAccDockerService_fullSpec` fails.

* master branch: https://travis-ci.org/terraform-providers/terraform-provider-docker/jobs/630041304
* pull request: https://travis-ci.org/terraform-providers/terraform-provider-docker/jobs/635266671?utm_medium=notification&utm_source=github_status

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

Rebuild CI.

### Important Factoids
Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? Custom version of OpenStack? Tight ACLs?

When I run `TestAccDockerService_fullSpec` at localhost, the test is passed.

```
$ make testacc_setup
$ TF_LOG=INFO TF_ACC=1 go test -v ./docker -run ^TestAccDockerService_fullSpec$ -timeout 360s
$ make testacc_cleanup
```

```
$ make testacc
```

My Environment

MacOS Mojave 10.14.6

```
$ go version
go version go1.13.5 darwin/amd64

$ docker version
Client: Docker Engine - Community
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        633a0ea
 Built:             Wed Nov 13 07:22:34 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.5
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.12
  Git commit:       633a0ea
  Built:            Wed Nov 13 07:29:19 2019
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- GH-1234

https://github.com/terraform-providers/terraform-provider-docker/pull/234#issuecomment-572841809

**Comment by suzuki-shunsuke** [comment]
@mavogel 

https://github.com/terraform-providers/terraform-provider-docker/pull/233

Maybe we should refer not `Spec.Endpoint` but `EndpointSpec`

https://travis-ci.org/terraform-providers/terraform-provider-docker/jobs/635539888#L1030-L1045

```
2020/01/11 01:09:18 [DEBUG] Docker service inspect: {
	"ID": "ufi9obypkm4sjl8arviukrmmr",
	"Version": {
		"Index": 15
	},
	"CreatedAt": "2020-01-11T01:09:18.797704555Z",
	"UpdatedAt": "2020-01-11T01:09:18.797704555Z",
	"Spec": {
		"Name": "tftest-service-basic",
		"Labels": {
			"servicelabel": "true"
		},
		"TaskTemplate": {
			"ContainerSpec": {
				"Image": "127.0.0.1:15000/tftest-service:v1@sha256:39425cb9c55596d21f3be8251ffe8bd7aea0fa26f3506e4be449dd9e0f8d881b",
				"Labels": {
					"foo": "bar"
				},
				"Command": [
					"ls"
				],
				"Args": [
					"-las"
				],
				"Hostname": "my-fancy-service",
				"Env": [
					"URI=/api-call?param1=value1",
					"MYFOO=BAR"
				],
				"Dir": "/root",
				"User": "root",
				"Groups": [
					"docker",
					"foogroup"
				],
				"Privileges": {
					"CredentialSpec": null,
					"SELinuxContext": {
						"Disable": true,
						"User": "user-label",
						"Role": "role-label",
						"Type": "type-label",
						"Level": "level-label"
					}
				},
				"StopSignal": "SIGTERM",
				"ReadOnly": true,
				"Mounts": [
					{
						"Type": "volume",
						"Source": "tftest-volume",
						"Target": "/mount/test",
						"ReadOnly": true,
						"VolumeOptions": {
							"NoCopy": true,
							"Labels": {
								"foo": "bar"
							},
							"DriverConfig": {
								"Name": "random-driver",
								"Options": {
									"op1": "val1"
								}
							}
						}
					}
				],
				"StopGracePeriod": 10000000000,
				"Healthcheck": {
					"Test": [
						"CMD",
						"curl",
						"-f",
						"localhost:8080/health"
					],
					"Interval": 5000000000,
					"Timeout": 2000000000,
					"Retries": 4
				},
				"Hosts": [
					"testhost:10.0.1.0"
				],
				"DNSConfig": {
					"Nameservers": [
						"8.8.8.8"
					],
					"Search": [
						"example.org"
					],
					"Options": [
						"timeout:3"
					]
				},
				"Secrets": [
					{
						"File": {
							"Name": "/secrets.json",
							"UID": "0",
							"GID": "0",
							"Mode": 777
						},
						"SecretID": "c8iq80harsivf29ksvhwd7fp6",
						"SecretName": "tftest-mysecret"
					}
				],
				"Configs": [
					{
						"File": {
							"Name": "/configs.json",
							"UID": "0",
							"GID": "0",
							"Mode": 292
						},
						"ConfigID": "eb63ikj36uaxeq0v0e8exkwgg",
						"ConfigName": "tftest-full-myconfig"
					}
				],
				"Isolation": "default"
			},
			"Resources": {
				"Limits": {
					"NanoCPUs": 1000000,
					"MemoryBytes": 536870912
				}
			},
			"RestartPolicy": {
				"Condition": "on-failure",
				"Delay": 3000000000,
				"MaxAttempts": 4,
				"Window": 10000000000
			},
			"Placement": {
				"Constraints": [
					"node.role==manager"
				],
				"Preferences": [
					{
						"Spread": {
							"SpreadDescriptor": "spread=node.role.manager"
						}
					}
				],
				"Platforms": [
					{
						"Architecture": "amd64",
						"OS": "linux"
					}
				]
			},
			"Networks": [
				{
					"Target": "sjtfg7m896apksac4w20y8ei5"
				}
			],
			"LogDriver": {
				"Name": "json-file",
				"Options": {
					"max-file": "3",
					"max-size": "10m"
				}
			},
			"ForceUpdate": 0,
			"Runtime": "container"
		},
		"Mode": {
			"Replicated": {
				"Replicas": 2
			}
		},
		"UpdateConfig": {
			"Parallelism": 2,
			"Delay": 10000000000,
			"FailureAction": "pause",
			"Monitor": 5000000000,
			"MaxFailureRatio": 0.1,
			"Order": "start-first"
		},
		"RollbackConfig": {
			"Parallelism": 2,
			"Delay": 5000000,
			"FailureAction": "pause",
			"Monitor": 36000000000000,
			"MaxFailureRatio": 0.9,
			"Order": "stop-first"
		},
		"EndpointSpec": {
			"Mode": "vip",
			"Ports": [
				{
					"Name": "random",
					"Protocol": "tcp",
					"TargetPort": 8080,
					"PublishedPort": 8080,
					"PublishMode": "ingress"
				}
			]
		}
	},
	"Endpoint": {
		"Spec": {}
	}
}
```

**Comment by mavogel** [comment]
> Maybe we should refer not Endpoint but Spec.EndpointSpec

Hmm yes basically reverting #233 . I don't know which `EndpointSpec` is the correct one to parse TBH as it seems to differ von env to env

**Comment by suzuki-shunsuke** [comment]
I agree.

**Comment by mavogel** [comment]
Maybe we could parse both and merge them together or wait until all fields are exposed, what I'm basically doing already for the `networks.`. What do you think about this?

**Comment by suzuki-shunsuke** [comment]
Thank you for your feedback.
I didn't know the feature `WaitForState`.

https://godoc.org/github.com/hashicorp/terraform/helper/resource#StateChangeConf.WaitForState

https://github.com/terraform-providers/terraform-provider-docker/blob/a7f6cc93009c4052f836c279e68ecd4a96ab238d/docker/resource_docker_network_funcs.go#L79
https://github.com/terraform-providers/terraform-provider-docker/blob/a7f6cc93009c4052f836c279e68ecd4a96ab238d/docker/resource_docker_network_funcs.go#L139-L182

I used this feature then CI passed!

https://github.com/terraform-providers/terraform-provider-docker/pull/238/commits/14bdb1c32ad266cc7e47f2b46d0f90895e136e63
https://travis-ci.org/terraform-providers/terraform-provider-docker/jobs/636116268

**Comment by mavogel** [comment]
Closed via #238

**Comment by suzuki-shunsuke** [comment]
@mavogel 
Unfortunately, CI is still broken due to other errors.

https://travis-ci.org/terraform-providers/terraform-provider-docker/builds/644754113?utm_source=github_status&utm_medium=notification

**Comment by suzuki-shunsuke** [comment]
> Error: Check failed: could not delete image 'sha256:d476e6d7f4df9834d68df2a461eed51355f2d86166b7846b9bc67d62aae2e279' after 6 retries

https://github.com/terraform-providers/terraform-provider-docker/blob/d8b0c139e5f9ec51b746ce399808e6442866d63b/docker/resource_docker_service_test.go#L1403-L1405

> image is being used by running container

It is failed to delete a Docker image because it is being used by running container.

**Comment by mavogel** [comment]
No worries @suzuki-shunsuke, I fixed it in ef47f0cbdb315bbe22ac0dcd291bc96098967b98 as mentioned in #238

**Comment by suzuki-shunsuke** [comment]
Oh, I got it.
Thank you for your fix! 👍 
https://github.com/terraform-providers/terraform-provider-docker/commits/master/docker/resource_docker_service_funcs.go
